Add vector search documentation (#9135)
opensearch-project/documentation-website: _vector-search/vector-search-techniques/painless-functions.md

---
layout: default
title: Painless extensions
nav_order: 25

> parent: Exact k-NN search with scoring script

Reviewer: with "a" scoring script?

grand_parent: Vector search techniques
has_children: false
has_math: true
redirect_from:
  - /search-plugins/knn/painless-functions/ 
---

# Painless scripting extensions

With the Painless scripting extensions, you can use k-NN distance functions directly in your Painless scripts to perform operations on `knn_vector` fields. Painless has a strict list of allowed functions and classes per context to ensure its scripts are secure. OpenSearch adds Painless Scripting extensions to a few of the distance functions used in [k-NN score script]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-score-script/), so you can use them to customize your k-NN workload.

> ## Get started with k-NN's Painless Scripting functions

Reviewer: ```suggestion
## Get started with k-NN Painless scripting functions
```

To use k-NN's Painless Scripting functions, first create an index with `knn_vector` fields like in [k-NN score script]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-score-script#getting-started-with-the-score-script-for-vectors). Once the index is created and you ingest some data, you can use the Painless extensions:

```json
GET my-knn-index-2/_search
{
  "size": 2,
  "query": {
    "script_score": {
      "query": {
        "bool": {
          "filter": {
            "term": {
              "color": "BLUE"
            }
          }
        }
      },
      "script": {
        "source": "1.0 + cosineSimilarity(params.query_value, doc[params.field])",
        "params": {
          "field": "my_vector",
          "query_value": [9.9, 9.9]
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

`field` needs to map to a `knn_vector` field, and `query_value` must be a floating-point array with the same dimension as `field`.

## Function types

The following table describes the Painless functions OpenSearch provides.

Function name | Function signature | Description

> :--- | :---

Reviewer: Line 57, last sentence: `divide by 0` errors?

`hamming` | `float hamming (float[] queryVector, doc['vector field'])` | This function calculates the Hamming distance between a given query vector and document vectors. The Hamming distance is the number of positions at which the corresponding elements are different. The shorter the distance, the more relevant the document is, so this example inverts the return value of the Hamming distance.

The `hamming` space type is supported for binary vectors in OpenSearch version 2.16 and later. For more information, see [Binary k-NN vectors]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-memory-optimized#binary-vectors).
{: .note}

## Constraints

1. If a document's `knn_vector` field has different dimensions than the query, the function throws an `IllegalArgumentException`.

2. If a vector field doesn't have a value, the function throws an `IllegalStateException`.

   You can avoid this situation by first checking if a document has a value in its field:

   ```
   "source": "doc[params.field].size() == 0 ? 0 : 1 / (1 + l2Squared(params.query_value, doc[params.field]))",
   ```

>    Because scores can only be positive, this script ranks documents with vector fields higher than those without.

Reviewer: ```suggestion
   Because scores can only be positive, this script ranks documents with vector fields higher than those without vector fields.
```

> With cosine similarity, it is not valid to pass a zero vector (`[0, 0, ...]`) as input. This is because the magnitude of such a vector is 0, which raises a `divide by 0` exception in the corresponding formula. Requests containing the zero vector will be rejected, and a corresponding exception will be thrown.

Reviewer: "With cosine similarity" => "When using cosine similarity"?

> With cosine similarity, it is not valid to pass a zero vector (`[0, 0, ...]`) as input. This is because the magnitude of such a vector is 0, which raises a `divide by 0` exception in the corresponding formula. Requests containing the zero vector will be rejected, and a corresponding exception will be thrown.

Author: ```suggestion
When using cosine similarity, it is not valid to pass a zero vector (`[0, 0, ...]`) as input. This is because the magnitude of such a vector is 0, which raises a `divide by 0` exception in the corresponding formula. Requests containing the zero vector will be rejected, and a corresponding exception will be thrown.
```
